### PR TITLE
Adds persistent collapse button for left side bar

### LINKED ISSUE #8684: Collapsable left nav
The left nav reduces important screen real estate and makes the UI feel very busy.

I propose adding a permanent  "collapse" button that keeps the left nave collapsed for all viewport sizes.

Thoughts?

### PR DESCRIPTION
**Pull request by jasonblanchard**
Fixes #8684 by adding a "collapse" and "expand" button on the left sidebar nav. State is stored in the `session` hash as suggested by @randx.

The purpose of this change is to allow for more screen real estate for larger viewports.

Here's what it looks like expanded

expanded

Here is is collapsed (ahhh, so refreshing :)

collapsed

The option is hidden on small viewports:

small

I'm not super in love with the button placement, so let me know if folks have other suggestions.

Note that I hit this bug https://github.com/activeadmin/activeadmin/issues/512 when testing the helper with the `session` hash. I used the fix suggested in that issue which required that I changed a few other tests. Everything appears to be working locally, but let me know if it causes other problems.

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://jasonblanchard-collapseable-nav.c68bfdc4cac4c32c41e1a922e73062ea.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://jasonblanchard-collapseable-nav.c68bfdc4cac4c32c41e1a922e73062ea.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

**Comment by jasonblanchard** [comment]
^ I see those failing tests, I don't think the current_user context matters here, I'll fix it

**Comment by jasonblanchard** [comment]
Ok, tests passing now. The persistent collapse now works for non-logged in visitors.

I've added a whole new controller for this, happy to put it somewhere else if it makes more sense.

File: app/views/nav/toggle_collapse_sidebar.js.erb [hunk 1]
```diff
@@ -0,0 +1,9 @@
+$('.collapse-nav').find('input[type="submit"]').prop('disabled', false).removeClass("disabled")
```
Comment (Razer6) [inline app/views/nav/toggle_collapse_sidebar.js.erb]: Actually, all view templates are implemented in HAML. Can you convert this to HAML?

File: spec/factories.rb [hunk 2]
```diff
@@ -1,5 +1,3 @@
-include ActionDispatch::TestProcess
```
Comment (Razer6) [inline spec/factories.rb]: Why was this removed?

File: spec/factories/notes.rb [hunk 3]
```diff
@@ -53,7 +53,7 @@
     end
 
     trait :with_attachment do
-      attachment { fixture_file_upload(Rails.root + "spec/fixtures/dk.png", "`/png") }
```
Comment (Razer6) [inline spec/factories/notes.rb]: How are these changes related to the navbar?

Comment (jasonblanchard) [inline spec/factories/notes.rb]: @Razer6 See my note about the test bug. This fix is recommended in https://github.com/activeadmin/activeadmin/issues/512

File: spec/factories.rb (hunk 2 above)
Comment (jasonblanchard) [inline spec/factories.rb]: https://github.com/activeadmin/activeadmin/issues/512

File: app/views/nav/toggle_collapse_sidebar.js.erb (hunk 1 above)
Comment (jasonblanchard) [inline app/views/nav/toggle_collapse_sidebar.js.erb]: Even though it's a `js.erb` file? I though haml was just for HTML templates, but let me know if I'm misunderstanding that. I don't usually use HAML...

File: spec/factories/notes.rb (hunk 3 above)
Comment (Razer6) [inline spec/factories/notes.rb]: Ok. Semaphore also seems to be happy with it :)

File: app/views/nav/toggle_collapse_sidebar.js.erb (hunk 1 above)
Comment (Razer6) [inline app/views/nav/toggle_collapse_sidebar.js.erb]: There is also a `js.haml`. Leave the decission to @randx

Comment (jasonblanchard) [inline app/views/nav/toggle_collapse_sidebar.js.erb]: K, easy enough to change. Didn't know that was a thing :)

File: app/views/nav/toggle_collapse_sidebar.js.haml [hunk 4]
```diff
@@ -0,0 +1,8 @@
+$('.collapse-nav').find('input[type="submit"]').prop('disabled', false).removeClass("disabled")
```
Comment (jasonblanchard) [inline app/views/nav/toggle_collapse_sidebar.js.haml]: @Razer6 Updated to `js.haml`

**Comment by randx** [comment]
I suggested to use cookies. (https://github.com/gitlabhq/gitlabhq/issues/8684#issuecomment-72160480) 
I think in this case we dont need new controller/routes and it will be only 2-3 files changed.

**Comment by jasonblanchard** [comment]
@randx ok, I can fix that implementation.

It's ok UI-wise?

**Comment by randx** [comment]
I think collapse button should be on bottom.

**Comment by jasonblanchard** [comment]
Ok, I'll try that

**Comment by jasonblanchard** [comment]
@randx Here's an updated implementation using cookies.

Also moved the expand/collapse button to the bottom:

screen shot 2015-02-02 at 8 50 51 pm

screen shot 2015-02-02 at 8 50 55 pm

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -44,6 +44,18 @@ def determine_layout
...
+      cookies.delete :collapsed_nav
+    else
+      cookies[:collapsed_nav] = true
+    end
+
+    respond_to do |format|
+      format.js
+    end
+  end
+
```
Comment (randx) [inline app/controllers/users_controller.rb]: why not just write cookies via JS?

File: app/assets/stylesheets/generic/sidebar.scss [hunk 6]
```diff
@@ -39,6 +39,42 @@
...
+    }
+  }
+}
+
+@media (max-width: 1198px) {
+  .collapse-nav {
+    display: none;
+  }
+}
+
```
Comment (randx) [inline app/assets/stylesheets/generic/sidebar.scss]: Navigation sidebar described in this css file - https://github.com/gitlabhq/gitlabhq/blob/master/app/assets/stylesheets/sections/nav_sidebar.scss

File: app/assets/stylesheets/sections/nav_sidebar.scss [hunk 7]
```diff
@@ -109,8 +109,8 @@
     width: $sidebar_width;
 
     .nav-sidebar {
-      margin-top: 20px;
-      position: fixed;
+      margin-top: 10px;
+      position: relative;
```
Comment (randx) [inline app/assets/stylesheets/sections/nav_sidebar.scss]: why this is necessary?

File: app/controllers/users_controller.rb (hunk 5 above)
Comment (jasonblanchard) [inline app/controllers/users_controller.rb]: I could. Would you prefer that implementation?

File: app/views/layouts/_collapse_button.html.haml [hunk 8]
```diff
@@ -0,0 +1,4 @@
+- if nav_menu_collapsed?
+  = link_to icon('plus-square'), toggle_collapse_sidebar_path, remote: true, :method => :post
```
Comment (randx) [inline app/views/layouts/_collapse_button.html.haml]: you can just write cookie with JS

**Comment by randx** [comment]
@jasonblanchard cool. I left some comments

File: app/controllers/users_controller.rb (hunk 5 above)
Comment (randx) [inline app/controllers/users_controller.rb]: @jasonblanchard yes. In this case we dont need controller, route and view

Comment (jasonblanchard) [inline app/controllers/users_controller.rb]: Will do.

File: app/assets/stylesheets/generic/sidebar.scss (hunk 6 above)
Comment (jasonblanchard) [inline app/assets/stylesheets/generic/sidebar.scss]: Great, I'll move this stuff there.

File: app/assets/stylesheets/sections/nav_sidebar.scss (hunk 7 above)
Comment (jasonblanchard) [inline app/assets/stylesheets/sections/nav_sidebar.scss]: This is how I was able to display the expand button below the nav elements.

Since the wrapping `.sidebar-wrapper` wrapper around `.nav-sidebar`, I don't think this needs to be, too.

**Comment by jasonblanchard** [comment]
@randx Updated with your comments. Implemented with a cookie in pure JS.
